File: backends/ruby-rails/app/usecase/invitation/interactor.rb
# frozen_string_literal: true
#
# 招待ユースケースの実装モジュール。
#
# @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>

module UseCase
  module Invitation
    # 招待に関するユースケースの実装クラスです。
    # @author Satoshi Nagashiba <satoshi.nagashiba@gmail.com>
    class Interactor
      # @param invitation_repo [Domain::Invitation::Repository] 招待リポジトリ
      # @param invitation_auth_repo [Domain::Invitation::AuthRepository] 招待認証キャッシュリポジトリ
      def initialize(invitation_repo, invitation_auth_repo)
        @invitation_repo      = invitation_repo
        @invitation_auth_repo = invitation_auth_repo
      end

      def current(role)
        entity = @invitation_repo.get_current_by_role(role)
        entity ? @invitation_repo.entity_to_vo(entity) : nil
      end

      def issue(role)
        entity = @invitation_repo.issue(role)
        @invitation_repo.entity_to_vo(entity)
      end

      def find_by_token(dto)
        entity = @invitation_repo.find_by_token(dto.token)
        raise "invitation_not_found" unless entity
        @invitation_auth_repo.store(entity.token, entity.role, 600)
        @invitation_repo.entity_to_vo(entity)
      end
    end
  end
end

File: backends/go-echo/internal/usecase/invitation/interactor.go
package invitation

import (
	dominvitation "authorization-go-echo/internal/domain/invitation"
	"authorization-go-echo/pkg/apperror"
)

type Interactor struct {
	repo     dominvitation.Repository
	authRepo dominvitation.AuthRepository
}

func NewInteractor(repo dominvitation.Repository, authRepo dominvitation.AuthRepository) *Interactor {
	return &Interactor{repo: repo, authRepo: authRepo}
}

func (uc *Interactor) Current(dto CurrentDto) (*dominvitation.Vo, error) {
	result, err := uc.repo.GetCurrentByRole(dto.Role)
	if err != nil {
		return nil, err
	}
	if result == nil {
		return nil, apperror.NotFound("invitation_not_found")
	}
	return result, nil
}

func (uc *Interactor) Issue(dto IssueDto) (*dominvitation.Vo, error) {
	return uc.repo.IssueByRole(dto.Role)
}

func (uc *Interactor) FindByToken(dto FindByTokenDto) (*dominvitation.Vo, error) {
	if dto.Token == "" {
		return nil, apperror.BadRequest("invitation_invalid")
	}
	result, err := uc.repo.FindByToken(dto.Token)
	if err != nil {
		return nil, err
	}
	if result == nil {
		return nil, apperror.BadRequest("invitation_invalid")
	}
	if err := uc.authRepo.Store(result.Token, result.Role, 600); err != nil {
		return nil, err
	}
	return result, nil
}
